### PR TITLE
Refactor Chain Service Block Processing API

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #1586 

---

# Description

**Write why you are making the changes in this pull request**

Our ChainService contains a big `ReceiveBlock` function which does too many things at once. It handles p2p, logic, db, state transitions, and more. We are in need of a cleaner, smaller API which can be exposed for other services to use bits and pieces. An example is in initial sync where we only need to apply state transitions and do not need p2p or other bloat when processing incoming blocks.

**Write a summary of the changes you are making**

This PR splits up the chain service's `ReceiveBlock` functions into a cleaner API. The new function is broken up into smaller functions which do the following

1. Verify a block passes pre-processing conditions
2. Save and broadcast the block via p2p to other peers
3. Apply the block state transition function and account for skip slots.
4. Process and cleanup any block operations, such as attestations and deposits, which would need to be either included or flushed from the beacon node's runtime.

### CODE REVIEW

File: beacon-chain/blockchain/block_processing.go [hunk 1]
```diff
@@ -18,77 +18,99 @@ import (
...
-//		# process slot with block
-//		state = slot_state_transition(state, block)
-//
-//		# check state root
-//      if block.state_root == hash(state):
-//			return state, error
-//		else:
-//			return nil, error  # or throw or whatever
-//
+// any block that is received from p2p layer or rpc. It performs the following actions:It checks the block to see
```
Comment (terenc3t) [inline beacon-chain/blockchain/block_processing.go]:
```suggestion
// any block that is received from p2p layer or rpc. It performs the following actions: It checks the block to see
```

File: beacon-chain/sync/initial-sync/service.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/blockchain"
+
```
Comment (terenc3t) [inline beacon-chain/sync/initial-sync/service.go]:
```suggestion
```

File: beacon-chain/rpc/service.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ import (
 	"net"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/blockchain"
+
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: space

File: beacon-chain/sync/initial-sync/service.go (hunk 2 above)
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/service.go]: space

File: beacon-chain/blockchain/block_processing.go [hunk 4]
```diff
@@ -106,13 +128,55 @@ func (c *ChainService) ReceiveBlock(ctx context.Context, block *pb.BeaconBlock)
...
+//   The node has processed its state up to slot, block.slot - 1.
+//   The Ethereum 1.0 block pointed to by the state.processed_pow_receipt_root has been processed and accepted.
+//   The node's local clock time is greater than or equal to state.genesis_time + block.slot * SECONDS_PER_SLOT.
+func (c *ChainService) VerifyBlockValidity(block *pb.BeaconBlock, beaconState *pb.BeaconState) error {
+	if block.Slot == params.BeaconConfig().GenesisSlot {
+		return fmt.Errorf("cannot process a genesis block: received block with slot %d",
+			block.Slot-params.BeaconConfig().GenesisSlot)
+	}
+	var powBlockFetcher func(ctx context.Context, hash common.Hash) (*gethTypes.Block, error)
+	if c.enablePOWChain {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/block_processing.go]: Can we remove `enablePOWChain`? Its a requirement now

**Pull request merged**